_source/_posts/2020-04-29-master-filestream-in-csharp-aspnetcore.md

```diff
@@ -0,0 +1,319 @@
+
+Now that we have content, it is time to use it.
+
+In order to have an endpoint that we can download our stream from, we need to set up a new controller. Make it an empty Controller, like below, and name it *FileController*.
+
+{% img blog/filestream-csharp/10-add-controller.png alt:"Add MVC Controller" width:"800" %}{: .center-image }
+
+```csharp
+[Authorize]
+public class FileController : ControllerBase
+{
+    IAmazonS3 _s3Client;
+
+    public FileController(IAmazonS3 s3Client)
+    {
+        _s3Client = s3Client;
+    }
+
+    public async Task Index()
+    {
+        //Our domain specific security logic would go here
+        var emailClaim = User.Claims.FirstOrDefault(x => x.Type == "email");
+        var email = emailClaim.Value;
+
+        if (email == "chase@cupcakefactory.net")
+        {
+            using (var fileStream = await _s3Client.GetObjectStreamAsync("your-bucket", "somefile", null))
+            {
+                this.Response.ContentType = "application/json";
+                this.Response.Headers.Add("Content-Disposition", "attachment; filename=\"filename.json\"");
+                this.Response.StatusCode = 200;
+                await fileStream.CopyToAsync(this.Response.Body);
+            }
+        }
+    }
+}
+```
+
+Let's break down what the code above is doing. The first thing we want to do is make sure no one that is logged in can access our controller. As Okta is already set up, all we need to do is add the **Authorize** decorator to the controller class.
+
+```csharp
+[Authorize]
+public class FileController : ControllerBase
+```
+
+Next, since the AWS SDK took care of loading our s3 client into our services, we can access it via dependency injection.
+
+```csharp
+IAmazonS3 _s3Client;
+public FileController(IAmazonS3 s3Client)
+{
+    _s3Client = s3Client;
+}
+```
+
+Okta gives us access to any information about our user that we ask for. This lets us have pretty robust logic for authorization. In my case, I'm filtering by email, but this could be by domain, role, age, or any other arbitrary claim that helps you determine authorization.
+
+```csharp
+var emailClaim = User.Claims.FirstOrDefault(x => x.Type == "email");
+var email = emailClaim.Value;
+
+if (email == "chase@cupcakefactory.net")
+```
+
+Lastly, because of .NET Core, we have direct access to the response body stream and can stream our bytes directly to the output without having to download the whole content first. We can also do any processing of the data before writing to the output stream the way we did in our first example.
+
+```csharp
+using (var fileStream = await _s3Client.GetObjectStreamAsync("your-bucket", "somefile", null))
+{
+    //Whatever your content type is
+    this.Response.ContentType = "application/json";
+    this.Response.Headers.Add("Content-Disposition", "attachment; filename=\"myfile.json\"");
+    this.Response.StatusCode = 200;
+    await fileStream.CopyToAsync(this.Response.Body);
+}
+```
+
+## Final thoughts
```

Comment:
I would combine this into the last section since "Final thoughts" isn't an H2 that's good for SEO. At the very least, capitalize Thoughts.

Reply:
Agreed. Will capitalize.